### PR TITLE
Add support for citations in PDFs/LaTeX

### LINKED ISSUE #18: Generalize to other writers (LatexWriter)
*Transferred from https://github.com/JuliaQuantumControl/QuantumCitations.jl/issues/7*

It might be better not to generate raw HTML for the bibliography. See the comments at https://github.com/JuliaDocs/Documenter.jl/issues/1162#issuecomment-658476560

### PR DESCRIPTION
**Pull request by goerz**
* Support for bibliographies in PDFs generate via LaTeX (`format=Documenter.LaTeX()`). Citations and references are rendered exactly as in the HTML version. Specifically, the support does not depend on `bibtex`/`biblatex` and supports any style (including custom styles).
* Functions `DocumenterCitations.set_latex_options` and `DocumenterCitations.reset_latex_options` to tweak the rendering of bibliographies in PDFs.

Internal changes:

* Any `@bibliography` block is now internally expanded into an internal `BibliographyNode` instead of a raw HTML node. This `BibliographyNode` can then be translated into the desired output format by `Documenter.HTMLWriter` or `Documenter.LaTeXWriter`. This is how support for bibliographies with `format=Documenter.LaTeX()` can be achieved.
* The routine `format_bibliography_reference` must now return a markdown string instead of an HTML string.

There is now a script `docs/makepdf.jl` that can be executed with `make pdf` to produce a PDF of the documentation for `DocumenterCitations`. This serves as a self-test and showcase. In future releases, we'll attach the PDF as an asset to the release. I didn't add a Github Workflow to produce the PDF, as setting up LaTeX and compiling the PDF would slow down the CI significantly. There are some tests, but they stop after generating the `.tex` file, which is much more efficient. For releases, I can produce the PDF locally on my laptop.

Closes #18

### CODE REVIEW

**Comment by goerz** [comment]
The documentation preview is at https://juliadocs.org/DocumenterCitations.jl/previews/PR46/

More importantly, the PDF version (generated with `make pdf`) is [DocumenterCitations.jl-1.2.1.pdf](https://github.com/JuliaDocs/DocumenterCitations.jl/files/12889986/DocumenterCitations.jl-1.2.1.pdf)

**Comment by goerz** [comment]
@Seelengrab Can you review this?

If I was your thesis advisor, I'd be totally fine with you handing in something in the format of the generated [PDF](https://github.com/JuliaDocs/DocumenterCitations.jl/files/12889986/DocumenterCitations.jl-1.2.1.pdf) 😉

Just let me know if this works for you, or if anything should be tweaked.

File: src/bibliography_node.jl [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+        list_tag = ol
+    end
+    html_list = list_tag()
+    for item in bibliography.items
+        anchor_key = isnothing(item.anchor_key) ? "" : "#$(item.anchor_key)"
+        html_reference = Documenter.HTMLWriter.domify(dctx, item.reference.children)
+        if bibliography.list_style == :dl
+            html_label = Documenter.HTMLWriter.domify(dctx, item.label.children)
+            push!(html_list.nodes, dt(html_label))
+            push!(html_list.nodes, dd(div["$(anchor_key)"](html_reference)))
```
Comment (Seelengrab) [inline src/bibliography_node.jl:72]: Aren't these `anchor_key` already strings at this point?

File: src/bibliography_node.jl [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+* `list_style`: One of `:dl`, `:ul`, `:ol`, cf. [`bib_html_list_style`](@ref)
+* `canonical`: Whether or not the references in the `@bibliography` block are
+  link targets.
+* `items`: A list of [`BibliographyItem`](@ref) objects, one for each reference
+in the block
+"""
+struct BibliographyNode <: Documenter.AbstractDocumenterBlock
+    list_style::Symbol  # one of :dl, :ul, :ol
+    canonical::Bool
+    items::Vector{BibliographyItem}
```
Comment (Seelengrab) [inline src/bibliography_node.jl:35]:
```suggestion
    list_style::Symbol  # one of :dl, :ul, :ol
    canonical::Bool
    items::Vector{BibliographyItem}
    function BibliographyNode(style, canon, items)
        style in (:dl, :ul, :ol) || throw(ArgumentError("Only :dl, :ul or :ol is allowed as a list style!"))
        new(style, canon, items)
    end
```

Better to check these assumptions than throw later on, if this is assumed throughout the code :)

File: src/formatting.jl [hunk 3]
```diff
@@ -208,9 +208,9 @@ end
 function format_published_in(entry; include_date=true)
     str = ""
     if entry.type == "article"
-        str *= entry.in.journal
+        str *= replace(entry.in.journal, " " => " ")  # non-breaking space
```
Comment (Seelengrab) [inline src/formatting.jl:211]: These two strings both contain only U+0020 (the regular space character) for me. I'd use the `\uxxxx` mode if a specific replacement is needed:

```julia-repl
julia> "\u01" |> first
'\x01': ASCII/Unicode U+0001 (category Cc: Other, control)
```

**Review by Seelengrab — CHANGES_REQUESTED** [review]
Just some minor things - I'll see how this works on my thesis in a few hours :)

**Comment by Seelengrab** [comment]
Something seems to be going wrong when linking down to the reference:

image

The relevant bibtex that got parsed:

```
@book{guntheroth2016optimized,
  title={Optimized C++: proven techniques for heightened performance},
  author={Guntheroth, Kurt},
  year={2016},
  publisher={" O'Reilly Media, Inc."}
}
```

And the TEX that got produced for this section:

```
This is an additional reference [\hyperlinkref{13874326461024448687}{1}].



{\raggedright% @bibliography

\hangindent=0.33in {\makebox[{\ifdim0.33in<\dimexpr\width+1ex\relax\dimexpr\width+1ex\relax\else0.33in\fi}][l]{[1]}}\hypertarget{13874326461024448687}{}K. Guntheroth. \emph{Optimized C++: proven techniques for heightened performance}. {\textquotedbl} O{\textquotesingle}Reilly Media, Inc.{\textquotedbl} (2016).

}% end @bibliography
```

The author name actually does contain `U+00A0` instead of a regular space, which may be related to my comment about space replacement from above? Not sure what's going wrong with the hyperlink though, might be something in my setup :thinking:

File: src/bibliography_node.jl (hunk 2 above)
Comment (goerz) [reply src/bibliography_node.jl:35]: Fair enough, although this is already checked earlier in the code. So the only way for a user to trigger that error would be to instantiate a `BibliographyNode` manually.

File: src/formatting.jl (hunk 3 above)
Comment (goerz) [reply src/formatting.jl:211]: Agreed: it's not a good idea to have invisible unicode in the code. I'll fix it. I was being lazy, and I have to figure out the code point for a non-breaking space. 

But when you say "These two strings both contain only the regular space character", that seems to be a bug in the Github UI! In the committed code, the second space is non-breaking, but I can confirm that if I copy and paste the above from the web UI into my text editor, both strings are regular spaces.

**Comment by goerz** [comment]
@Seelengrab Which platform are you running on? Is this a Windows problem? That `U+00A0` is supposed to be there, but it certainly shouldn't render like that. I'll look into whether I can replace "\u00a0" with "~" somewhere in the LaTeX writer. Are you having problems with any other unicode symbols?

**Comment by Seelengrab** [comment]
I'm on linux - my setup is a bit weird though, since I need a preamble from my university and that messes with things a bit :joy: That may be the cause of the wonky `ifundefined` too.. I *think* have it set up correctly for unicode support now (e.g. I can use ß and similar just fine, and other unicode like BenchmarkTools.jl output also looks great). So I think this is specific to these lower control characters?

**Comment by goerz** [comment]
>  I'll look into whether I can replace "\u00a0" with "~" somewhere in the LaTeX writer.

That will require a patch to Documenter, but is probably a good idea

**Comment by goerz** [comment]
> since I need a preamble from my university and that messes with things a bit 

That seems the likely cause for the `ifundefined`. Not sure if there's anything I can do about that.

Until we find a way to replace `\u00a0` with `~`, you can probably do that manually in your build script: Let Documenter only produce the `.tex` file (`platform="none"`), and search and replace `\u00a0` with the `~` (with `sed` or something), and then compile with `lualatex -shell-escape`

**Comment by Seelengrab** [comment]
Sounds good! I'm excited for this, since the reference really does look MUCH nicer now than my previous cobbled together attempt :D Thank you very much!

File: src/bibliography_node.jl (hunk 1 above)
Comment (goerz) [reply src/bibliography_node.jl:76]: Yes, thanks! Probably should be `anchor_id`, though, since it has the `#` prepended (making it an HTML id element).

**Comment by goerz** [comment]
I opened https://github.com/JuliaDocs/Documenter.jl/pull/2300 about the problem with the non-breaking space.

My best guess is that this is a font problem on your system. It's not that surprising that the LaTeX font you may be using doesn't handle `\u00A0` correctly, since `~` is really the expected way to have non-breaking spaces in LaTeX.

I'm going to see how quickly @mortenpi gets around to looking at https://github.com/JuliaDocs/Documenter.jl/pull/2300, and maybe keep this PR open until a new version of Documenter is released with that patch. That's because I have to adjust my tests that check the output of the `LaTeXrWiter`, depending on what version of Documenter is installed.

**Comment by Seelengrab** [comment]
Yeah, dealing with unicode & fonts in LaTeX is a bit of a PitA, unfortunately.. It's doubly bad when it shows up like this :/ I'll see if I can fiddle around a bit with the fonts and work out what (if anything) is going wrong on my end.

**Comment by goerz** [comment]
> I'm going to see how quickly @mortenpi gets around to looking at https://github.com/JuliaDocs/Documenter.jl/pull/2300,

I think he was saying on Slack that he was going to be traveling for a few weeks, so maybe I'll just merge later today (unless @Seelengrab has more input) and move on to fixing #22 next. I can always adjust the test on `master` when there's a new release of Documenter.

**Pull request merged**